lc0/src/neural/network_blas_cl_common.h

```diff
@@ -0,0 +1,119 @@
+    Copyright (C) 2018 The LCZero Authors
+
+    Leela Zero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Leela Zero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Leela Zero.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/* A table of variable conversions from lczero/Network.cpp to lc0/network.h
+
+get_input_channels()     -> kInputPlanes
+Network::initialize::channels -> ConvBlock.biases.size()
+conv_weights             -> ConvBlock.weights
+conv_biases              -> ConvBlock.biases
+batchnorm_means          -> ConvBlock.bn_means
+batchnorm_stddivs        -> ConvBlock.bn_stddivs
+conv_pol_w               -> Weights.policy.weights
+conv_pol_b               -> Weights.policy.biases
+bn_pol_w1                -> Weights.policy.bn_means,
+bn_pol_w2                -> Weights.policy.bn_stddivs // ??? not very useful oldnames
+NUM_POLICY_INPUT_PLANES  -> Weights.policy.bn_means.size()
+ip_pol_w                 -> Weights.ip_pol_w
+ip_pol_b                 -> Weights.ip_pol_b
+get_num_output_policy()  -> Weights.ip_pol_b.size()
+conv_val_w               -> Weights.value.weights
+conv_val_b               -> Weights.value.biases
+bn_val_w1                -> Weights.value.bn_means
+bn_val_w2                -> Weights.value.bn_stddivs
+NUM_VALUE_INPUT_PLANES   -> Weights.value.bn_means.size()
+{ip1_val_w,ip1_val_b,ip2_val_w,ip2_val_b} -> Weights.{}
+NUM_VALUE_CHANNELS       -> Weights.ip1_val_b.size()
+
+*/
+
+#pragma once
+
+#include "factory.h" // network.h, optionsdict.h
+
+namespace lczero {
+
+class BlasCLNetwork;
+
+class BlasCLNetworkComputation : public NetworkComputation {
+
+  BlasCLNetworkComputation(BlasCLNetwork* network) : network(network) {}
+
+  virtual void AddInput(InputPlanes&& input) override {
+    inputs.push_back(input);
+  }
+
+  virtual int GetBatchSize() const override {
+    return inputs.size();
+  }
+
+  virtual float GetQVal(int sample) const override {
+    return output_values[sample];
+  }
+  
+  virtual float GetPVal(int sample, int move_id) const override {
+    return output_policies[sample][move_id];
+  }
+
+  virtual void ComputeBlocking() override;
+
+ private:
+  std::vector<InputPlanes> inputs;
+  std::vector<float>   output_values;
+  std::vector<std::vector<float>> output_policies;
+  BlasCLNetwork* network;
+};
+
+class BlasCLNetwork : public Network {
+ public:
+  BlasCLNetwork(const Weights& weights, const OptionsDict& options)
+    : weights(weights), options(options) {
+    initialize();
```

Comment:
I have some aversion to putting too much code in a constructor. Something about future flexibility and maintainability, I think. I've had many cases in various codebases before where I had to factor out functionality from a constructor because it turned out to be more generally useful, and of course constructors aren't arbitrarily callable.

Not that I expect any such here for the relatively straightforward use of neural networks, but you never know, and the habit is deep-seated in me. (Possible future OpenCL batching might be an example where the constructor gets more complicated? I'm not sure.)

I still prefer to leave it a separate function, but I'd compromise on moving the constructor definition out of the header to maintain a clean, legible declaration.

Reply:
Ok, I'm fine with leaving it as it is.
(Although I disagree. Constructor is just a function, but it has clear semantic which disallows incorrect usage. Much code in any function is bad, it indeed makes sense to split large functions into several).

Reply:
Also when initially designing this, I had been thinking there would be some inheritance/delegating to super, though it didn't turn out that way. I'll move the contstructor definition and remove initialize.